docs/proposals/202104_native_support_of_multitenancy.md

```diff
@@ -0,0 +1,83 @@
+---
+title: Tracking the tenant’s data usages by supporting multi-tenancy natively
+type: proposal
+menu: proposals
+status: proposed
+owner: Abhishek357
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/3835 (*: Flexible Multi-tenancy Data Model)
+* https://github.com/thanos-io/thanos/issues/3834 (Multi tenant rule evaluation)
+* https://github.com/thanos-io/thanos/issues/3822 (Ability to require tenant in queries)
+* https://github.com/thanos-io/thanos/issues/3819 (Thanos receive per tenant limits)
+* https://github.com/thanos-io/thanos/issues/3820 (Zone aware replication)
+* https://github.com/thanos-io/thanos/issues/3572 (Track data usage across tenant / system)
+* https://github.com/thanos-io/thanos/issues/1318 (*: Support multi-tenancy on object storage using custom prefix)
+* https://github.com/thanos-io/thanos/pull/3289 (receiver/compact/store/sidecar: Added a provision in the bucket config to add a prefix)
+
+### Summary
+
+We want to introduce a special tenant label for identifying a tenant and make each component aware of it and then finally track the data usages across the tenants. The receiver is on the safe side as it has a tenant-specific label to uniquely identify tenants.
+
+All the components implementing either the Store API (Sidecar, Ruler, Store Gateway, Receiver) or Query API (Thanos querier frontend and Thanos querier) will be affected. Currently, the ruler is exposing the store API but since there is an ongoing LFX project (to make the ruler stateless), it will upload the data to the receiver itself and it might no longer expose Store API.
+
+### Motivation
+
+* We actually don't have any way to track the resource usages of the query(like error rates, memory allocation, CPU usages along with the queries) and tenant-wise metrics.
+* Every component should allow passing a special label that allows APIs that identify tenants.
+* Currently, we use external labels to achieve multi-tenancy. These labels are sufficient for the sidecar (single-tenant per sidecar i.e hard tenant) but for other components (where we will have multiple tenants per component i.e soft tenancy) these external labels are not sufficient.
+
+### Goals
+
+1. Introduce a mechanism to identify tenants to support multi-tenancy use cases.
+2. Isolate all the incoming query requests.
```

Comment:
Nice! Is query API the only goal? If yes let's change this to:

```suggestion
2. Isolate the incoming read and write API requests across tenants if needed.
```

If only query then let's make sure to put `Metadata, Exemplars` APIs  etc to ### Non-Goals section

Reply:
yes, for now, let's focus on query API.
other opinions are welcomed

Reply:
We can make a note about store API(only the querier part) that we are focussing on?

Reply:
Well it would be nice to put it in other APIs too, but we can extend that later. Just important @Abhishek357 to decide now (: